### ISSUE TITLE
Fix SRVConnector to deal with bad nameservers.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2777]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)| @ralphm reported|
|-|-|
|Trac ID|trac#2777|
|Type|defect|
|Created|2007-08-08 09:38:41Z|

Some nameservers reply with a `NXDOMAIN` when there are just no SRV records for that particular domain. We should then try to just use the fallback (plain host name and default port for that service). If the domain really doesn't exist, the subsequent resolving of the host will fail anyway.

<details><summary>Searchable metadata</summary>

```
trac-id__2777 2777
type__defect defect
reporter__ralphm ralphm
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1186565921000000 1186565921000000
changetime__1187945105000000 1187945105000000
version__None None
owner__ 
cc__ralphm cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_1"></a>|@ralphm commented|
|-|-|

Started [branches/srv-nxdomain-2777,](branches/srv-nxdomain-2777,) but `twisted.names.srvconnect` doesn't have any existing tests yet, so I'd need to dive deeper in this.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_2"></a>|@ralphm removed owner|
|-|-|

Fixed in that branch, also addresses [#1908](https://github.com/twisted/twisted/issues/1908).

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @ralphm|
|-|-|

Hum... I guess this is a huge improvement :). I know nothing about SRV, but your changes looks reasonable. As there were no tests, it's hard to tell if you break something.

My only comment: update copyright of srvconnect to 2007, and restrict the one of test_srvconnect to 2007, then please merge.

Also, I think we should open a ticket for general cleanups of srvconnect, like full test coverage and removal of all these crazy asserts.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_4"></a>|@ralphm set status to `closed`|
|-|-|

(In [21159]) Make SRVConnector work with bad results and NXDOMAIN responses.

Author: ralphm
Reviewer: therve
Fixes [#1908](https://github.com/twisted/twisted/issues/1908), [#2777](https://github.com/twisted/twisted/issues/2777).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
